catalog/app/containers/Search/model.ts

```diff
@@ -698,50 +706,328 @@ export function useNextPagePackagesQuery(after: string) {
+    buckets: model.state.buckets,
+    filter,
+  })
+
+  return GQL.fold(query, {
     data: ({ searchPackages: r }) => {
       switch (r.__typename) {
         case 'EmptySearchResultSet':
-          return NO_FACETS
         case 'InvalidInput':
-          return NO_FACETS
+          return children(AvailableFiltersState.Empty())
         case 'PackagesSearchResultSet':
-          return r.stats.userMeta
+          return React.createElement(AvailablePackagesMetaFiltersReady, {
+            facets: r.stats.userMeta,
+            truncated: r.stats.userMetaTruncated,
+            children,
+          })
         default:
           assertNever(r)
       }
     },
-    fetching: () => NO_FACETS,
-    error: () => NO_FACETS,
+    fetching: () => children(AvailableFiltersState.Loading()),
+    error: () => children(AvailableFiltersState.Empty()),
   })
-  const activated = model.state.userMetaFilters.filters
-  const activatedPaths = React.useMemo(() => Array.from(activated.keys()), [activated])
+}
+
+function AvailablePackagesMetaFiltersReady({
+  facets,
+  truncated,
+  children,
+}: RenderProps<AvailableFiltersStateInstance> & {
+  facets: readonly PackageUserMetaFacet[]
+  truncated: boolean
+}) {
+  const { filters } = useSearchUIModelContext(ResultType.QuiltPackage).state
+    .userMetaFilters
 
   const available = React.useMemo(
-    () => all.filter((f) => !activatedPaths.includes(f.path)),
-    [all, activatedPaths],
+    () => facets.filter((f) => !filters.has(f.path)),
+    [facets, filters],
   )
 
-  return { all, activated, activatedPaths, available, fetching: query.fetching }
+  // server-side filtering required
+  if (truncated) {
+    return React.createElement(AvailablePackagesMetaFiltersServerFilter, {
+      available,
+      children,
+    })
+  }
+
+  // client-side filtering required
+  if (available.length >= FACETS_FILTER_THRESHOLD) {
+    return React.createElement(AvailablePackagesMetaFiltersClientFilter, {
+      available,
+      children,
+    })
+  }
+
+  if (!available.length) {
+    return children(AvailableFiltersState.Empty())
+  }
+
+  // no filtering required
+  return React.createElement(AvailablePackagesMetaFiltersGroup, {
+    state: AvailableFiltersState.Ready({
+      filtering: FacetsFilteringState.Disabled(),
+      facets: {
+        available,
+        visible: EMPTY_FACET_TREE,
+        hidden: EMPTY_FACET_TREE,
+      },
+      fetching: false,
+    }),
+    children,
+  })
+}
+
+function AvailablePackagesMetaFiltersServerFilter({
+  children,
+  available,
+}: RenderProps<AvailableFiltersStateInstance> & {
+  available: readonly PackageUserMetaFacet[]
+}) {
+  const [path, setPath] = React.useState('')
+  let pathNorm = path.trim().toLowerCase()
+  // add wildcards to use substring matching by default
+  if (pathNorm && !pathNorm.includes('*') && !pathNorm.includes('?')) {
+    pathNorm = `*${pathNorm}*`
+  }
+  const [pathDebounced] = useDebounce(pathNorm, 500)
+
+  if (pathDebounced) {
```

Comment:
I think, because of this condition, there is a delay when you start typing, or when you clear the input

Reply:
yep, and this is kinda expected